src/implementation.cpp

```diff
@@ -1685,6 +1685,63 @@ simdutf_warn_unused size_t convert_utf16_to_utf8(const char16_t *buf,
   return convert_utf16le_to_utf8(buf, len, utf8_buffer);
   #endif
 }
+
+simdutf_warn_unused size_t
+convert_utf16_to_utf8_safe(const char16_t *buf, size_t len, char *utf8_output,
+                           size_t utf8_len) noexcept {
+  const auto start{utf8_output};
+  // We might be able to go faster by first scanning the input buffer to
+  // determine how many char16_t characters we can read without exceeding the
+  // utf8_len. This is a one-pass algorithm that has the benefit of not
+  // requiring a first pass to determine the length.
+  while (true) {
+    // The worst case for convert_utf16_to_utf8 is when you go from 1 char16_t
+    // to 3 characters of UTF-8. So we can read at most utf8_len / 3 char16_t
+    // characters.
+    auto read_len = std::min(len, utf8_len / 3);
+    if (read_len <= 16) {
+      break;
+    }
+    if (read_len < len) {
+      // printf("checking value %zx\n", uint16_t(buf[read_len - 1]));
+      //  If we have a high surrogate at the end of the buffer, we need to
+      //  either read one more char16_t or backtrack.
+      if (buf[read_len - 1] >= 0xD800 && buf[read_len - 1] <= 0xDBFF) {
```

Comment:
Rather than 2 lookups, this might be faster:

```
auto last_char = buf[read_len - 1];
if ((last_char & 0xFC00) == 0xD800) {
  read_len--;
}
```

Reply:
Done.